### PR TITLE
Add MCM support in LQ (new device API)

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Native-execution of circuits with mid-measurements is enabled via the dynamic-one-shot transform. To support the transform, a device must implement methods to apply MidMeasureMP.

MidMeasureMP is a special kind of operation where a sample is drawn from the state vector probability distribution and a projector is applied to the state vector accordingly. It may also support options like reset and post-selection.

The most straightforward way to deal with the requirements is to enhance the state vector class StateVectorLQubit.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
#621 
Add mid circuit measurements to lightning [sc-58448]
Lightning Qubit implement MidMeasureMP method [sc-57704]

### CODE REVIEW

File: pennylane_lightning/lightning_qubit/_measurements.py [hunk 1]
```diff
@@ -72,7 +70,7 @@ class LightningMeasurements:
 
     def __init__(
         self,
-        qubit_state: LightningStateVector,
+        qubit_state,
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit/_measurements.py:73]: Why remove that?

File: tests/test_native_mcm.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+from conftest import LightningDevice as ld
+from conftest import device_name
+from flaky import flaky
+from pennylane._device import DeviceError
+
+if not ld._CPP_BINARY_AVAILABLE:
+    pytest.skip("No binary module found. Skipping.", allow_module_level=True)
+
+if device_name != "lightning.qubit":
+    pytest.skip(f"Native MCM not supported in device {device_name}.", allow_module_level=True)
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:30]:
```suggestion
if not LightningDevice._new_API:
    pytest.skip("Exclusive tests for new device API. Skipping.", allow_module_level=True)

if not LightningDevice._CPP_BINARY_AVAILABLE:  # pylint: disable=protected-access
    pytest.skip("No binary module found. Skipping.", allow_module_level=True)
```
Let's keep it in a standard way with the other tests in this folder. My idea with this is to make it as straightforward as possible for someone to migrate our other devices.

File: tests/test_native_mcm.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+# limitations under the License.
+"""Tests for default qubit preprocessing."""
+from functools import reduce
+from typing import Sequence
+
+import numpy as np
+import pennylane as qml
+import pytest
+from conftest import LightningDevice as ld
+from conftest import device_name
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:22]:
```suggestion
from conftest import device_name, LightningDevice
```

File: tests/test_native_mcm.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        func(*params)
+
+
+@flaky(max_runs=5)
+@pytest.mark.parametrize("shots", [5000, [5000, 5001]])
+@pytest.mark.parametrize("postselect", [None, 0, 1])
+@pytest.mark.parametrize("reset", [False, True])
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
+def test_simple_mcm(shots, postselect, reset, measure_f):
+    """Tests that LightningQubit handles a circuit with a single mid-circuit measurement and a
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:171]:
```suggestion
    """Tests that LightningDevice handles a circuit with a single mid-circuit measurement and a
```

File: tests/test_native_mcm.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+        validate_measurements(measure_f, shots, results1, results2)
+
+
+@flaky(max_runs=5)
+@pytest.mark.parametrize("shots", [1000, [1000, 1001]])
+@pytest.mark.parametrize("postselect", [None, 0, 1])
+@pytest.mark.parametrize("reset", [False, True])
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
+def test_single_mcm_single_measure_mcm(shots, postselect, reset, measure_f):
+    """Tests that LightningQubit handles a circuit with a single mid-circuit measurement and a
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:201]:
```suggestion
    """Tests that LightningDevice handles a circuit with a single mid-circuit measurement and a
```

File: tests/test_native_mcm.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+
+
+@flaky(max_runs=5)
+@pytest.mark.parametrize("shots", [3000, [3000, 3001]])
+@pytest.mark.parametrize("postselect", [None, 0, 1])
+@pytest.mark.parametrize("reset", [False, True])
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.probs, qml.sample])
+@pytest.mark.parametrize("wires", [[0], [0, 1]])
+def test_single_mcm_single_measure_wires(shots, postselect, reset, measure_f, wires):
+    """Tests that LightningQubit handles a circuit with a single mid-circuit measurement and a
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:279]:
```suggestion
    """Tests that LightningDevice handles a circuit with a single mid-circuit measurement and a
```

File: tests/test_native_mcm.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        validate_measurements(measure_f, shots, results1, results2)
+
+
+@flaky(max_runs=5)
+@pytest.mark.parametrize("shots", [5000])
+@pytest.mark.parametrize("postselect", [None, 0, 1])
+@pytest.mark.parametrize("reset", [False, True])
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
+def test_single_mcm_multiple_measurements(shots, postselect, reset, measure_f):
+    """Tests that LightningQubit handles a circuit with a single mid-circuit measurement with reset
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:308]:
```suggestion
    """Tests that LightningDevice handles a circuit with a single mid-circuit measurement with reset
```

File: tests/test_native_mcm.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+            validate_measurements(measure_f, shots, r1, r2)
+
+
+@flaky(max_runs=5)
+@pytest.mark.parametrize("shots", [5000, [5000, 5001]])
+@pytest.mark.parametrize("postselect", [None, 0, 1])
+@pytest.mark.parametrize("reset", [False, True])
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.sample, qml.var])
+def test_composite_mcm_measure_composite_mcm(shots, postselect, reset, measure_f):
+    """Tests that LightningQubit handles a circuit with a composite mid-circuit measurement and a
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:338]:
```suggestion
    """Tests that LightningDevice handles a circuit with a composite mid-circuit measurement and a
```

File: tests/test_native_mcm.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        validate_measurements(measure_f, shots, results1, results2)
+
+
+@flaky(max_runs=5)
+@pytest.mark.parametrize("shots", [5000, [5000, 5001]])
+@pytest.mark.parametrize("postselect", [None, 0, 1])
+@pytest.mark.parametrize("reset", [False, True])
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.expval, qml.probs, qml.sample, qml.var])
+def test_composite_mcm_single_measure_obs(shots, postselect, reset, measure_f):
+    """Tests that LightningQubit handles a circuit with a composite mid-circuit measurement and a
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:371]:
```suggestion
    """Tests that LightningDevice handles a circuit with a composite mid-circuit measurement and a
```

File: tests/test_native_mcm.py [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+        validate_measurements(measure_f, shots, results1, results2)
+
+
+@flaky(max_runs=5)
+@pytest.mark.parametrize("shots", [5000, [5000, 5001]])
+@pytest.mark.parametrize("postselect", [None, 0, 1])
+@pytest.mark.parametrize("reset", [False, True])
+@pytest.mark.parametrize("measure_f", [qml.counts, qml.probs, qml.sample])
+def test_composite_mcm_measure_value_list(shots, postselect, reset, measure_f):
+    """Tests that LightningQubit handles a circuit with a composite mid-circuit measurement and a
```
Comment (AmintorDusko) [inline tests/test_native_mcm.py:401]:
```suggestion
    """Tests that LightningDevice handles a circuit with a composite mid-circuit measurement and a
```

File: pennylane_lightning/lightning_qubit/_measurements.py (hunk 1 above)
Comment (vincentmr) [reply pennylane_lightning/lightning_qubit/_measurements.py:73]: `LightningStateVector` is solely imported for this type annotation, but `LightningMeasurements` is required in `_state_vector.py` so removing this avoids a cyclic dependency.

Comment (AmintorDusko) [reply pennylane_lightning/lightning_qubit/_measurements.py:73]: Makes sense! Thank you for that!

File: .github/workflows/tests_linux.yml [hunk 11]
```diff
@@ -33,7 +33,7 @@ jobs:
       matrix:
         os: [ubuntu-22.04]
         pl_backend: ["lightning_qubit"]
-    timeout-minutes: 30
+    timeout-minutes: 60
```
Comment (maliasadi) [inline .github/workflows/tests_linux.yml:36]: Just for my knowledge, why did you need 60min timeout? :thinking:

File: Makefile [hunk 12]
```diff
@@ -6,8 +6,10 @@ PL_BACKEND ?= "$(if $(backend:-=),$(backend),lightning_qubit)"
 
 ifdef check
     CHECK := --check
+	ICHECK := --check --diff
```
Comment (maliasadi) [inline Makefile:9]:
```suggestion
    CHECK := --check
    ICHECK := --check --diff
```

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 13]
```diff
@@ -72,13 +79,6 @@
         create_ops_listC128,
     )
 
-    # pylint: disable=import-error, no-name-in-module, ungrouped-imports
-    from pennylane_lightning.core._serialize import (
-        QuantumScriptSerializer,
-        global_phase_diagonal,
-    )
-    from pennylane_lightning.core._version import __version__
-
```
Comment (maliasadi) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:81]: these headers changes again :sweat_smile:

File: pennylane_lightning/lightning_qubit/_state_vector.py [hunk 14]
```diff
@@ -246,14 +250,36 @@ def _apply_lightning_controlled(self, operation):
                 False,
             )
 
-    def _apply_lightning(self, operations):
+    def _apply_lightning_midmeasure(self, operation: MidMeasureMP, mid_measurements: dict):
```
Comment (maliasadi) [inline pennylane_lightning/lightning_qubit/_state_vector.py:253]: could it be `None`?

```suggestion
    def _apply_lightning_midmeasure(self, operation: MidMeasureMP, mid_measurements: dict = None):
```

File: tests/test_native_mcm.py [hunk 15]
```diff
@@ -0,0 +1,423 @@
...
+        sh1, sh2 = results1.shape[0], results2.shape[0]
+        assert np.allclose(sh1, sh2, rtol=20, atol=0.2)
+        assert results1.ndim == results2.ndim
+        if results2.ndim > 1:
+            assert results1.shape[1] == results2.shape[1]
+        np.allclose(np.sum(results1), np.sum(results2), rtol=20, atol=0.2)
+
+
+def validate_expval(shots, results1, results2):
+    """Compares two expval, probs or var.
```
Comment (maliasadi) [inline tests/test_native_mcm.py:78]:
```suggestion
    """Compares two expval.
```

File: tests/test_native_mcm.py [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+        return
+
+    if func is qml.sample:
+        validate_samples(shots, results1, results2)
+        return
+
+    validate_expval(shots, results1, results2)
+
+
+def test_all_invalid_shots_circuit():
```
Comment (maliasadi) [inline tests/test_native_mcm.py:111]:
```suggestion
def test_all_invalid_shots_circuit():
    """Test all invalid cases.
    """

```

Comment (maliasadi) [reply tests/test_native_mcm.py:111]: is this currect?
```suggestion
def test_all_invalid_shots_circuit():
    """Test all invalid cases: var and probs.
    """

```

File: tests/test_native_mcm.py [hunk 17]
```diff
@@ -0,0 +1,423 @@
...
+    res1 = qml.QNode(circuit_mcm, dq)()
+    res2 = qml.QNode(circuit_mcm, dev)(shots=10)
+    for r1, r2 in zip(res1, res2):
+        if isinstance(r1, Sequence):
+            assert len(r1) == len(r2)
+        assert np.all(np.isnan(r1))
+        assert np.all(np.isnan(r2))
+
+
+def test_unsupported_measurement():
```
Comment (maliasadi) [inline tests/test_native_mcm.py:146]:
```suggestion
def test_unsupported_measurement():
    """Test unsupported ``qml.classical_shadow`` measurement on ``lightning.qubit``.
    """

```

**Review by maliasadi — APPROVED** [review]
Oops, I ran into meetings this morning without having the chance to submit my review comments...

File: .github/workflows/tests_linux.yml (hunk 11 above)
Comment (vincentmr) [reply .github/workflows/tests_linux.yml:36]: 30 minutes is just a bit short now that we have native mcm tests. The one-by-one shot implementation is unfortunately rather inefficient for more than ~1000 shots and we do need more than that to get even 1 or 2-digit matching on all parametrized tests.

File: Makefile (hunk 12 above)
Comment (vincentmr) [reply Makefile:9]: Weird :(

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py (hunk 13 above)
Comment (vincentmr) [reply pennylane_lightning/lightning_kokkos/lightning_kokkos.py:81]: Hopefully fixed when #647 is merged.

File: pennylane_lightning/lightning_qubit/_state_vector.py (hunk 14 above)
Comment (vincentmr) [reply pennylane_lightning/lightning_qubit/_state_vector.py:253]: I'd rather not because the sole purpose is to apply `measure` and this should return an MCM.

File: tests/test_native_mcm.py (hunk 15 above)
Comment (vincentmr) [reply tests/test_native_mcm.py:78]: The docstring is OK, I'll change the function name.

File: pennylane_lightning/lightning_qubit/_state_vector.py (hunk 14 above)
Comment (maliasadi) [reply pennylane_lightning/lightning_qubit/_state_vector.py:253]: hmm.. right, right! then let's update the docs.

File: pennylane_lightning/lightning_qubit/_state_vector.py [hunk 18]
```diff
@@ -246,14 +250,36 @@ def _apply_lightning_controlled(self, operation):
...
+        """Execute a MidMeasureMP operation and return the sample in mid_measurements.
+        Args:
+            operation (~pennylane.operation.Operation): mid-circuit measurement
+            mid_measurements (None, dict): Dictionary of mid-circuit measurements
```
Comment (maliasadi) [inline pennylane_lightning/lightning_qubit/_state_vector.py:257]:
```suggestion
            mid_measurements (dict): Dictionary of mid-circuit measurements
```
